filebeat/input/filestream/input.go

```diff
@@ -168,7 +184,160 @@ func (inp *filestream) Run(
+		return false, fmt.Errorf("cannot unpack cursor from '%s' to read offset: %w",
+			fs.newPath,
+			err)
+	}
+
+	graceTimer := time.NewTimer(inp.deleterConfig.GracePeriod)
+	checkIntervalTicker := time.NewTicker(inp.scannerCheckInterval)
+	// Wait for the grace period or for the context to be cancelled
+LOOP:
+	for {
+		select {
+		case <-ctx.Cancelation.Done():
+			return false, ctx.Cancelation.Err()
+		case <-checkIntervalTicker.C:
+			stat, err := os.Stat(fs.newPath)
+			if err != nil {
+				// Return the error and cause the harvester to close
+				return false, fmt.Errorf("cannot stat '%s': %w", fs.newPath, err)
+			}
+
+			if stat.Size() != st.Offset {
+				// The file has changed, close the harvester
+				// without deleting the file
+				logger.Debugf("cancelling deletion of '%s', size has changed from %d to %d",
+					fs.newPath,
+					st.Offset,
+					stat.Size())
+				return false, nil
+			}
+		case <-graceTimer.C:
+			break LOOP
+		}
+	}
+
+	stat, err := os.Stat(fs.newPath)
+	if err != nil {
+		return false, fmt.Errorf("cannot stat '%s': %w", fs.newPath, err)
+	}
+
+	// The file has been written to, close the harvester so the filewatcher
+	// can start a new one
+	if stat.Size() != st.Offset {
+		logger.Debugf("'%s' was updated, won't remove. Closing harvester", fs.newPath)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (inp *filestream) deleteFile(
+	ctx input.Context,
+	logger *logp.Logger,
+	cursor loginp.Cursor,
+	fs fileSource,
+) error {
+	// Wait until the resource is finished.
+	// The resource is finished when all events are acknowledged by the output.
+	// If it is not finished, this usually means one of two things:
+	//   - The output is down
+	//   - Filebeat is experiencing back pressure
+	// Either way, we wait until all events have been published.
+	if !cursor.Finished() {
+		logger.Debugf(
+			"not all events from '%s' have been published, "+
+				"closing harvester",
+			fs.newPath)
+		// normal operation of the harvester, the file was closed.
+		return nil
+	}
+	logger.Infof(
+		"all events from '%s' have been published, waiting for %s grace period",
+		fs.newPath, inp.deleterConfig.GracePeriod.String())
+
+	canRemove, err := inp.waitGracePeriod(ctx, logger, cursor, fs)
+	if err != nil {
+		return err
+	}
+
+	if !canRemove {
+		return nil
+	}
+
+	if err := os.Remove(fs.newPath); err != nil {
+		// The first try at removing the file failed,
+		// retry with a constant backoff
+		lastErr := err
+
+		ticker := time.NewTicker(2 * time.Second)
+		defer ticker.Stop()
+
+		retries := 0
+		for retries < 5 {
+			logger.Errorf(
+				"could not remove '%s', retrying in 2s. Error: %s",
+				fs.newPath,
+				lastErr,
+			)
+
+			select {
+			case <-ctx.Cancelation.Done():
+				return ctx.Cancelation.Err()
+
+			case <-ticker.C:
+				retries++
+				lastErr = os.Remove(fs.newPath)
+				if lastErr == nil {
+					logger.Infof("'%s' removed", fs.newPath)
+					return nil
+				}
+			}
+		}
+
+		return fmt.Errorf(
```

Comment:
Will the harvester pick up the file again and try to delete it after the grace period if we hit this case?

I think this is the behavior we want, if some other process on Windows has the file open for an hour we should still eventually delete it.

We perhaps want an explicit test case for this where deletion fails during backoff.

Reply:
You might want to consider making the backoff interval and max configurable but not document them just to making testing this easier.

Reply:
>  Will the harvester pick up the file again and try to delete it after the grace period if we hit this case?

Any error will cause the harvester to close, then on the next file scan the file is picked up again, if it still exists.

We don't hold any "state" regarding the delete outside the harvester that is running. When the harvester closes the file is treated like any other: the file scan will discover it, the harvester will start, the file will eventually close, then the delete code will run.